### DISCUSSION TITLE
Best practice for albums having Various Artists?

### DISCUSSION THREAD

**Discussion by Lucki** [nukeop/nuclear#1668] [Q&A] [answered]
What is the best practice to handle albums having Various Artists?

I added from the MusicBrainz search some albums for nostalgia, but the matched tracks are less than underwhelming.
* Need for Speed Underground 2
* Need for Speed Most Wanted (2005)

Now the problems I have are:
* I can't see the original artists in the track list so I have to remember or search the original artists to adjust false tracks
* The spinning thing on each track in the currently playing list resets the popup, so I can't adjust any false tracks until the scanning completes
* Often the correct track isn't available in the dropdown and this is the most annyoing part

Specific examples from the albums above:
* I Need Speed from Capone, the second track from NFSU2 is not available to select
* I Do from Chingy, the third track from NFSU2 is not available to select
* Do Ya Thang from TI presents The P$C, the second track from NFSMW is not available to select, only a strange short 2 minute version
* I am Rock from Rock, the third track from NFSMW is not available to select

At this point I gave up, am I doing something wrong? All mentioned tracks are available on YT and the first search entries when searching for the track and album name without the artist. [1](https://www.youtube.com/results?search_query=i+need+speed+need+for+speed+underground+2) [2](https://www.youtube.com/results?search_query=I+do+need+for+speed+underground+2) [3](https://www.youtube.com/results?search_query=do+ya+thang+need+for+speed+most+wanted) [4](https://www.youtube.com/results?search_query=i+am+rock+need+for+speed+most+wanted)

Bonus bug?
The track numbering seems scrambled with multiple entries having the same number in the NFSU2 album.

**Comment by nukeop** [comment]
Nuclear uses the artist's name and the track's title when looking up streams. So if the album's artist is "Various Artists" and the title is "Track 1", it will search for "Various Artsits - Track 1", and try to match a stream to that.

**Reply by Lucki** [reply]
This clarifies why the matches don't fit at all.

Judging from the lack of displayed track artists, I imagine this is a problem on the metadata provider side of things?
The artists are available for individual tracks on MusicBrainz, but I see `"artist": {"name": "Various Artists"},` added to each track in the config file.

I also noticed the stream selection dropdown actually has an entry to type stuff into, but it didn't eat the direct YouTube link -doing that for all tracks would be cumbersome anyway.

I guess I'll craft different issues out of this whole thing.

**Reply by nukeop** [reply]
You can correct the matched streams using stream verification: https://nukeop.gitbook.io/nuclear/user-manual/stream-verification

**Reply by Lucki** [reply]
Sure, but I wasn't even able to select and play the correct stream, so verifying isn't an option.

To be clear: I am unable to play the correct streams for (some tracks in) the example album. I can't select them from the context menu drop down because the search term is too generic to have the correct result as an option to choose.

**Reply by nukeop** [reply]
Some metadata providers can keep the particular artist for each track, so it's possible to improve this in the future and use the correct search terms.

**Reply by Lucki** [reply]
I played a bit around. It doesn't seem too hard, actually.
This test patch already works for both Spotify and MusicBrainz.

<details>
<summary>patch</summary>

~~~ patch
diff --git a/packages/app/app/components/AlbumView/index.tsx b/packages/app/app/components/AlbumView/index.tsx
index 0a22cb63..88970e06 100644
--- a/packages/app/app/components/AlbumView/index.tsx
+++ b/packages/app/app/components/AlbumView/index.tsx
@@ -41,6 +41,20 @@ export const AlbumView: React.FC<AlbumViewProps> = ({
   const displayPlaylistCreationDialog = () => setIsCreatePlaylistDialogOpen(true);
   const hidePlaylistCreationDialog = () => setIsCreatePlaylistDialogOpen(false);
 
+  const displayArtistColumn = () => {
+    if (!album.artist || !album.tracklist) {
+      return false;
+    }
+
+    for (const track of album.tracklist) {
+      if (album.artist !== track.artist) {
+        return true;
+      }
+    }
+
+    return false;
+  };
+
   const release_date: Date = new Date(album.year);
   return <div className={styles.album_view_container}>
     <Dimmer.Dimmable>
@@ -178,7 +192,7 @@ export const AlbumView: React.FC<AlbumViewProps> = ({
               tracks={album.tracklist}
               displayDeleteButton={false}
               displayThumbnail={false}
-              displayArtist={false}
+              displayArtist={displayArtistColumn()}
               displayAlbum={false}
             />
           </div>
diff --git a/packages/app/app/containers/AlbumViewContainer/hooks.ts b/packages/app/app/containers/AlbumViewContainer/hooks.ts
index f7654adc..9623164a 100644
--- a/packages/app/app/containers/AlbumViewContainer/hooks.ts
+++ b/packages/app/app/containers/AlbumViewContainer/hooks.ts
@@ -39,10 +39,13 @@ export const useAlbumViewProps = () => {
       thumbnail: album.coverImage,
       duration: parseInt(track.duration) !== track.duration
         ? stringDurationToSeconds(track.duration)
-        : track.duration,
-      artist: {
-        name: album.artist
-      }
+        : track.duration
+      // This throws a react error, don't know why - both are strings
+      // The error message claims react expects an array but making this an array
+      // prevents the error as well as the artist being shown
+      // artist: {
+      //   name: track.artist
+      // }
     }));
   }
 
@@ -70,7 +73,7 @@ export const useAlbumViewProps = () => {
   const addAlbumToQueue = useCallback(async () => {
     await album?.tracklist.forEach(async track => {
       dispatch(QueueActions.addToQueue({
-        artist: album?.artist,
+        artist: track.artist,
         name: track.title,
         thumbnail: album.coverImage,
         streams: []
diff --git a/packages/core/src/plugins/meta/musicbrainz.ts b/packages/core/src/plugins/meta/musicbrainz.ts
index bf9dde61..5b4d7c7c 100644
--- a/packages/core/src/plugins/meta/musicbrainz.ts
+++ b/packages/core/src/plugins/meta/musicbrainz.ts
@@ -144,7 +144,7 @@ class MusicbrainzMetaProvider extends MetaProvider {
       tracklist: _.flatMap(releaseDetails.media, medium => _.map(medium.tracks, track => {
         const newtrack = new Track();
         newtrack.ids[SearchResultsSource.Musicbrainz] = track.id;
-        newtrack.artist = artistName;
+        newtrack.artist = track['artist-credit'][0].name;
         newtrack.title = track.title;
         newtrack.duration = Math.ceil(track.length/1000);
         newtrack.position = track.position;
diff --git a/packages/core/src/plugins/meta/spotify.ts b/packages/core/src/plugins/meta/spotify.ts
index 478485f5..94a88783 100644
--- a/packages/core/src/plugins/meta/spotify.ts
+++ b/packages/core/src/plugins/meta/spotify.ts
@@ -167,7 +167,7 @@ export class SpotifyMetaProvider extends MetaProvider {
       year: result.release_date,
       tracklist: result.tracks.items.map(track => new Track({
         title: track.name,
-        artist: result.artists[0].name,
+        artist: track.artists[0].name,
         duration: track.duration_ms/1000,
         position: track.track_number,
         thumbnail: thumb
diff --git a/packages/core/src/rest/Musicbrainz.ts b/packages/core/src/rest/Musicbrainz.ts
index e2ed294c..f3e4f75f 100644
--- a/packages/core/src/rest/Musicbrainz.ts
+++ b/packages/core/src/rest/Musicbrainz.ts
@@ -62,7 +62,7 @@ const getReleaseGroupDetails = (releaseGroupId: string): Promise<MusicbrainzRele
 
 const getReleaseDetails = (releaseId: string): Promise<MusicbrainzRelease> =>
   new Promise((resolve, reject) => {
-    nb.release(releaseId, { inc: 'artists+recordings+genres' }, (err, response) => {
+    nb.release(releaseId, { inc: 'artists+recordings+genres+artist-credits' }, (err, response) => {
       err ? reject(err) : resolve(response);
     });
   });
diff --git a/packages/core/src/rest/Musicbrainz.types.ts b/packages/core/src/rest/Musicbrainz.types.ts
index daa20508..e7a65401 100644
--- a/packages/core/src/rest/Musicbrainz.types.ts
+++ b/packages/core/src/rest/Musicbrainz.types.ts
@@ -9,12 +9,17 @@ export type MusicbrainzReleaseGroup = {
   releases: MusicbrainzRelease[];
 }
 
+export type MusicBrainzArtistCredit = {
+  name: string;
+}
+
 export type MusicbrainzTrack = {
   id: string;
   length: number;
   number: string;
   position: number;
   title: string;
+  'artist-credit': MusicBrainzArtistCredit[];
 };
 
 export type MusicbrainzMedia = {
~~~

</details>

**Comment by Lucki** [comment, marked as answer]
At this point in time it's not possible yet.

There's work on the way to get this going: #1681 
I also separated the additional issues into their dedicated things: #1671 #1672
